src/core/opamStd.ml

```diff
@@ -995,19 +995,27 @@ module OpamSys = struct
       (fun f -> try f () with _ -> ())
       !registered_at_exit
 
-  let is_cygwin_variant =
+  let get_windows_executable_variant =
     if Sys.win32 then
       let results = Hashtbl.create 17 in
-      let requires_cygwin name =
+      let requires_cygwin_or_msys2 name =
         let cmd = Printf.sprintf "cygcheck \"%s\"" name in
         let ((c, _, _) as process) = Unix.open_process_full cmd (Unix.environment ()) in
         let rec f a =
           match input_line c with
           | x ->
-              if OpamString.ends_with ~suffix:"cygwin1.dll" (String.trim x) then
+              let tx = String.trim x in
+              if OpamString.ends_with ~suffix:"cygwin1.dll" tx then
                 if OpamString.starts_with ~prefix:"  " x then
                   f `Cygwin
-                else if a <> `Cygwin then
+                else if a <> `Cygwin && a <> `Msys2 then
+                  f `CygLinked
+                else
+                  f a
+              else if OpamString.ends_with ~suffix:"msys-2.0.dll" tx then
+                if OpamString.starts_with ~prefix:"  " x then
+                  f `Msys2
+                else if a <> `Cygwin && a <> `Msys2 then
                   f `CygLinked
```

Comment:
```suggestion
                else if a = `Native then
                  f (`Tainted `Cygwin)
                else
                  f a
              else if OpamString.ends_with ~suffix:"msys-2.0.dll" tx then
                if OpamString.starts_with ~prefix:"  " x then
                  f `Msys2
                else if a = `Native then
                  f (`Tainted `Msys2)
```

Reply:
It's OK not to handle the very bizarre case of an executable which has managed to link with both msys2 and cygwin-based DLLs!